### PR TITLE
HDDS-8447. Datanodes should not process container deletes for failed volumes.

### PR DESCRIPTION
**Pull request by sadanand48**
## What changes were proposed in this pull request?

1. Check if the volume is in the failed volume set before deleting a container. If volume is in failed volume set eventually the failedVolumeListener will pick it up and remove it from memory.
2. Trigger an on-demand volume scan if there is an IO error while moving the container dir to tmp location during container deletion.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-8447

## How was this patch tested?
Unit test

### CODE REVIEW

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java [hunk 1]
```diff
@@ -1382,6 +1391,8 @@ private void deleteInternal(Container container, boolean force)
         if (!success) {
           LOG.error("Failed to move container under " +
               hddsVolume.getDeleteServiceDirPath());
+          // Trigger a volume scan as rename failed
+          StorageVolumeUtil.onFailure(container.getContainerData().getVolume());
```
Comment (ashishkumar50) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:1395]: Before this we also check `container.isEmpty()` and `logBlocksIfNonZero(container)` to make sure container is really empty or not, which will throw `IOException` as well if there is problem in volume as we are accessing disk and rocksdb here. So we need to do trigger volume scan at that place too. 
Also https://github.com/apache/ozone/pull/4655 which will refactor this method and will probably merge in a day, I think we should handle this change on top of that.

**Review by ashishkumar50 — COMMENTED** [review]
Hi Sadanand, Thanks for working on this Please find comment inline.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java (hunk 1 above)
Comment (sadanand48) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java]: Thanks @ashishkumar50 , updated the change.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java [hunk 2]
```diff
@@ -1352,9 +1356,11 @@ private void deleteInternal(Container container, boolean force)
       // empty as a defensive check.
       LOG.error("Could not determine if the container {} is empty",
           container.getContainerData().getContainerID(), e);
-      throw new StorageContainerException("Could not determine if container "
-          + container.getContainerData().getContainerID() +
-          " is empty", DELETE_ON_NON_EMPTY_CONTAINER);
+      String errorMsg =
+          "Could not determine if the container " + container.getContainerData()
+              .getContainerID() + "is empty";
+      triggerVolumeScanAndThrowException(container, errorMsg,
```
Comment (krishnaasawa1) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:1362]: What is difference here in overall flow, if we detect here Volume faulty on scan or not?  if volume is faulty will container will be removed  and DN will report success?

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java [hunk 3]
```diff
@@ -1363,4 +1369,12 @@ private void deleteInternal(Container container, boolean force)
     container.getContainerData().setState(State.DELETED);
     sendICR(container);
   }
+
+  private void triggerVolumeScanAndThrowException(Container container,
```
[the remark this answers is not in the document]
Comment (sadanand48) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:1373]: [StorageVolumeChecker](https://github.com/apache/ozone/blob/7a73a7a37721f3f793fe69becf5fad6f22b3a43e/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/StorageVolumeChecker.java#L258) handles this part , It schedules scans for these volumes and if this called multiple times for the same volume then it rejects the call as there is already a scan scheduled for that volume.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java (hunk 2 above)
Comment (sadanand48) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:1362]: If a volume is found faulty in the scan , it is marked as failed and then the failedVolumeListener will pick it up and remove containers that are present in that volume from the containerSet in memory and send an FCR . SCM will identify these as missing containers and trigger replication elsewhere .

Comment (krishnaasawa1) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:1362]: okay  so a single container delete failure here will result in volume scan. But if volume is found faulty, all containers within that will be removed and then FCR will be sent.

**Comment by ashishkumar50** [comment]
@errose28, Can you please look at this once.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java (hunk 2 above)
Comment (sadanand48) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:1362]: Yes right , that is how it should behave as containers in faulty volumes cannot be accessed.

**Comment by errose28** [comment]
I don't see code or tests checking if the volume has already failed when the container delete is being processed. Did I miss something here?

**Comment by sadanand48** [comment]
> I don't see code or tests checking if the volume has already failed when the container delete is being processed. Did I miss something here?

Thanks @errose28 for taking a look . Updated the patch, this was present in the first version of the patch, it somehow  got missed during rebase

**Review by errose28 — APPROVED** [review]
Thanks for the fix @sadanand48. One minor comment but otherwise LGTM.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java [hunk 4]
```diff
@@ -1352,9 +1366,11 @@ private void deleteInternal(Container container, boolean force)
...
       LOG.error("Could not determine if the container {} is empty",
           container.getContainerData().getContainerID(), e);
-      throw new StorageContainerException("Could not determine if container "
-          + container.getContainerData().getContainerID() +
-          " is empty", DELETE_ON_NON_EMPTY_CONTAINER);
+      String errorMsg =
+          "Could not determine if the container " + container.getContainerData()
+              .getContainerID() + "is empty";
+      triggerVolumeScanAndThrowException(container, errorMsg,
+          DELETE_ON_NON_EMPTY_CONTAINER);
```
[the remark this answers is not in the document]
Comment (sadanand48) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java]: Done.

**Comment by ashishkumar50** [comment]
@sadanand48, Thanks for updating patch, LGTM +1.

**Review by ashishkumar50 — APPROVED** [review]

**Comment by sadanand48** [comment]
Thanks @errose28 , @ashishkumar50  & @krishnaasawa1  for the reviews.

**Pull request merged**